### PR TITLE
[13.0][ENH] base_tier_validation, add escalation option

### PR DESCRIPTION
**Pull request by kittiu**
Add new feature: **Escalation**
- Allow Escalation in tier definition
- For that tier review, user has option to also Escalate
- Escalation means validate current tier and create extra tier for another reviewer

Pending issue:
- On escalated tier, when user make comment, it will set on the review. Because, to get the reviews, current logic needs definition first, I think it is not valid and require fixing too https://github.com/OCA/server-ux/pull/198

Depends on this fix:
- [X] https://github.com/OCA/server-ux/pull/198

Following is the example, we have 3 tiers approval by sequence for Purchase https://github.com/OCA/purchase-workflow/pull/967, where step 2 will allow user2 to Escalate to step 2.1. And only if step 2.1 is done, user 3 can proceed with step 3.

### CODE REVIEW

File: base_tier_validation/models/tier_review.py
Comment (kittiu) [inline base_tier_validation/models/tier_review.py:52]: kittiu: change to float because escalated tier will be insertion, i.e., 1.1, 2.1 etc.

File: base_tier_validation/models/tier_review.py
Comment (kittiu) [inline base_tier_validation/models/tier_review.py:12]: kittiu: I need to use compuate store=True, readonly=False of v13, because the extra escalated tier will not have definition. But still need these fields value.

**Comment by kittiu** [comment]
@pedrobaeza if you could share your opinion on Escalation feature that would be awesome. This feature is quite varied from user I ask, but here I try my best to make it simple and conform with existing.

**Comment by pedrobaeza** [comment]
Functionally seeing the GIFs (thanks for them), then only thing I see weird is that the step 2 is logged as approved instead of a new status "Escalated".

**Comment by kittiu** [comment]
Good point. I will add the new status "Escalated". Will be like this,

image

**Comment by kittiu** [comment]
Ready for review.

**Review by HviorForgeFlow — COMMENTED** [review]
Functional Review: :+1:
Code Review: I would rename escalate wizard to tier validation escalate wizard

**Comment by kittiu** [comment]
@HviorForgeFlow done changed model name from `escalate.wizard` to `validation.escalate.wizard`

**Comment by HviorForgeFlow** [comment]
I would prefer tier.validation.escalate.wizard sorry by the nitpicking

**Review by HviorForgeFlow — APPROVED** [review]

**Review by ps-tubtim — APPROVED** [review]
The code and functional test are LGTM.

**Comment by kittiu** [comment]
> I would prefer tier.validation.escalate.wizard sorry by the nitpicking

Done :+1:

**Comment by nmoturi** [comment]
Hi Kitti,

Is the base tier escalation enhancement includes the other enhancement related to server action & autovalidation..?

Thanks,
Naran

**Comment by kittiu** [comment]
> Hi Kitti,
> 
> Is the base tier escalation enhancement includes the other enhancement related to server action & autovalidation..?
> 
> Thanks,
> Naran

Nopl. You need to test them separately for now. Until both get merged.

**Comment by kittiu** [comment]
@HviorForgeFlow can this get merged? Thank you!

**Review by LoisRForgeFlow — COMMENTED** [review]
This is starting to get more and more complex and I feel I don't have use cases for some of this features while this changes can impact existing installations. Would it be possible to extract this kind of extended features to a separate module? then we can add a section in the settings wizard to activate extension on tier validations. Also, this is particular feature (escalation) it is not really documented (not even a help string in the has_escalation field) to help users to understand how it is going to behave.

I think if we can extract features in smaller modules and improve functional documentation, we will improve UX a lot.

**Comment by LoisRForgeFlow** [comment]
Also, if you separate in different modules it will be easier to mantain because the migrations will be more granular and the merges on changes faster because the impact will be limited to a module which is installed in less instances (you can even be maintainer of this subfeatures and control the release of changes).

**Comment by kittiu** [comment]
@LoisRForgeFlow yes, sure. I think this one can. This is your module, I want you to have the final say for things like this :) Will notify you after it is done.

Note: the thing is I do have many things I wanted to make base_teir_validation works more flexible, but I also aware about the complexity added into it. Your guidance would be much appreciated. https://github.com/OCA/server-ux/issues/192

**Comment by kittiu** [comment]
@LoisRForgeFlow https://github.com/OCA/server-ux/pull/220

Should we split this one too https://github.com/OCA/server-ux/pull/207 (to base_tier_validation_action)? I am ok, as it will be inline altogether.

**Comment by LoisRForgeFlow** [comment]
@kittiu yes, if you don't mind I rather make base_tier_validation_action also separate.

**Pull request closed without merging**
